### ISSUE TITLE
Support SIP over WebSockets (rfc7118)

### ISSUE THREAD

**Issue by Lekensteyn** [SIPp/sipp#124]
Originally requested in http://sourceforge.net/p/sipp/feature-requests/33/ for draft-ietf-sipcore-sip-websocket, now actually available as [RFC 7118][1].

This would be useful for testing real-world scenarios which operate over WSS.

Has anyone been working on this?

 [1]: https://tools.ietf.org/html/rfc7118

**Comment by wdoekes** [comment]
I don't think anyone has. Feel free to start working on it. Thanks!

**Comment by hpaskalev** [comment]
If no one is working on it I think of giving it a try.
I have some experience with WS already.
Or maybe I could take a stab at synchronous audio and video play.

**Comment by Lekensteyn** [comment]
The only change I have so far is:
https://github.com/CodeYellowBV/sipp/tree/websockets
https://github.com/CodeYellowBV/sipp/commit/8fbef73018fd3022bb94ccf377216650f3c3c2db

As the commit says, use an existing library or write one ourselves? As for debugging the traffic, use the 1.99.x development version of Wireshark for improved Websockets dissection.

**Comment by hpaskalev** [comment]
I can't recall standalone library that provides WS. I know WebRTC and reSIProcate have their implementations built-in. Best approach would probably be to look at the RFC and some reference implementation, and then write one ourselves. WS are relatively simple.

**Comment by Lekensteyn** [comment]
I found at least two libraries last time I checked (back in March), but those were either too bloated or looked too fragile. WS as a transport is simple, but integrating it in sipp can be more work. It should have a comparable complexity to SSL over TCP though. (btw, I am mainly interested in testing WSS instead of unencrypted WS).

**Comment by moises-silva** [comment]
@Lekensteyn did you look at libwebsockets.org? from briefly looking at the sources looks decent enough to me

**Comment by hpaskalev** [comment]
Well, I think of giving both WS and WSS a try. We might collaborate on this one. I think I should have something in the following days. Btw, I think of doing the thing from scratch.

**Comment by hpaskalev** [comment]
I must agree with @moises-silva - at first glance `llibwebsockets` sure looks good enough.

**Comment by Lekensteyn** [comment]
libwebsockets looks OK. Their commit logs look a it strange though with [improperly applied patches][1]. License is LGPL 2.1 btw and it v1.2.2 packaged since Debian Jessie (and Ubuntu 14.04).

 [1]: http://git.libwebsockets.org/cgi-bin/cgit/libwebsockets/commit/?id=edcef320e5641f6051d8347c388acd3be1d2ace6

**Comment by Lekensteyn** [comment]
@hpaskalev Have you made progress on this?

**Comment by hpaskalev** [comment]
No, I haven't started working on this one yet.
I've been busy with other things lately, but I plan to get onto it these days. Are you working on it ?

**Comment by hotshot47** [comment]
I am too looking for sipp with ws and wss transport as used extensively for testing our application.
Mean while any idea of other available UA supporting websocket apart from js based browser pluggins

**Comment by Lekensteyn** [comment]
@hpaskalev No time for it unfortunately, so haven't started with it yet :(

**Comment by hpaskalev** [comment]
Well, I just started working on it. :)

**Comment by andrewvmail** [comment]
+1

**Comment by Tanzeel-inline** [comment]
@hotshot47 Were you able to find any application for testing over wss?

Also, I think C++ have mature websocket libraries to integrate sipp over wss

**Comment by PadScratchJohns** [comment]
Bumping this one for WS/WSS support.
